File: crates/j-law-python/tests/test_withholding_tax.py
"""所得税法第204条第1項に基づく源泉徴収税額計算のテスト。"""

from __future__ import annotations

import datetime
import json
import pathlib

import pytest

from j_law_python.withholding_tax import (
    WithholdingTaxCategory,
    calc_withholding_tax,
)

_FIXTURE_PATH = (
    pathlib.Path(__file__).resolve().parents[3] / "tests" / "fixtures" / "withholding_tax.json"
)
_FIXTURES = json.loads(_FIXTURE_PATH.read_text(encoding="utf-8"))


@pytest.mark.parametrize("case", _FIXTURES["withholding_tax"], ids=lambda c: c["id"])
def test_withholding_tax(case):
    inp = case["input"]
    exp = case["expected"]
    date = datetime.date.fromisoformat(inp["date"])

    result = calc_withholding_tax(
        inp["payment_amount"],
        date,
        inp["category"],
        is_submission_prize=inp["is_submission_prize"],
        separated_consumption_tax_amount=inp["separated_consumption_tax_amount"],
    )

    assert result.taxable_payment_amount == exp["taxable_payment_amount"]
    assert result.tax_amount == exp["tax_amount"]
    assert result.net_payment_amount == exp["net_payment_amount"]
    assert result.submission_prize_exempted is exp["submission_prize_exempted"]


class TestLanguageSpecific:
    def test_out_of_range_date(self):
        with pytest.raises(ValueError):
            calc_withholding_tax(
                100_000,
                datetime.date(2012, 12, 31),
                WithholdingTaxCategory.MANUSCRIPT_AND_LECTURE,
            )

    def test_invalid_date_type(self):
        with pytest.raises(TypeError):
            calc_withholding_tax(100_000, "2026-01-01", "professional_fee")

    def test_breakdown_fields(self):
        result = calc_withholding_tax(
            1_500_000,
            datetime.date(2026, 1, 1),
            "professional_fee",
        )
        assert len(result.breakdown) == 2
        assert result.breakdown[0]["label"] != ""

    def test_repr(self):
        result = calc_withholding_tax(
            100_000,
            datetime.date(2026, 1, 1),
            "manuscript_and_lecture",
        )
        assert "WithholdingTaxResult" in repr(result)

File: crates/j-law-ruby/test/test_withholding_tax.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
require "date"
require "j_law_ruby"

class TestWithholdingTax < Minitest::Test
  FIXTURES = JSON.parse(File.read(File.join(__dir__, "../../../tests/fixtures/withholding_tax.json")))

  FIXTURES["withholding_tax"].each do |tc|
    define_method("test_#{tc['id']}") do
      inp = tc["input"]
      exp = tc["expected"]

      result = JLawRuby::WithholdingTax.calc_withholding_tax(
        inp["payment_amount"],
        Date.parse(inp["date"]),
        inp["category"],
        is_submission_prize: inp["is_submission_prize"],
        separated_consumption_tax_amount: inp["separated_consumption_tax_amount"]
      )

      assert_equal exp["taxable_payment_amount"], result.taxable_payment_amount
      assert_equal exp["tax_amount"], result.tax_amount
      assert_equal exp["net_payment_amount"], result.net_payment_amount
      assert_equal exp["submission_prize_exempted"], result.submission_prize_exempted?
    end
  end

  def test_out_of_range_date
    assert_raises(RuntimeError) do
      JLawRuby::WithholdingTax.calc_withholding_tax(
        100_000,
        Date.new(2012, 12, 31),
        :manuscript_and_lecture
      )
    end
  end

  def test_breakdown_fields
    result = JLawRuby::WithholdingTax.calc_withholding_tax(
      1_500_000,
      Date.new(2026, 1, 1),
      :professional_fee
    )
    assert_equal 2, result.breakdown.length
    refute_empty result.breakdown.first[:label]
  end

  def test_inspect
    result = JLawRuby::WithholdingTax.calc_withholding_tax(
      100_000,
      Date.new(2026, 1, 1),
      :manuscript_and_lecture
    )
    assert_match(/WithholdingTaxResult/, result.inspect)
  end

  def test_type_error_invalid_date
    assert_raises(TypeError) do
      JLawRuby::WithholdingTax.calc_withholding_tax(100_000, "2026-01-01", :professional_fee)
    end
  end
end
